crates/aggchain-proof-core/src/keccak/digest.rs

```
@@ -0,0 +1,131 @@
+        }
+
+        Ok(())
+    }
+}
+
+impl Digest {
+    pub const ZERO: Digest = Digest([0u8; 32]);
+
+    pub fn as_slice(&self) -> &[u8] {
+        self.0.as_slice()
+    }
+
+    pub fn as_bytes(&self) -> &[u8; 32] {
+        &self.0
+    }
+}
+
+impl From<[u8; 32]> for Digest {
+    fn from(bytes: [u8; 32]) -> Self {
+        Self(bytes)
+    }
+}
+
+impl fmt::LowerHex for Digest {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if f.alternate() {
+            write!(f, "0x")?;
+        }
+
+        for byte in &self.0 {
+            write!(f, "{:02x}", byte)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<'de> Deserialize<'de> for Digest {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        if deserializer.is_human_readable() {
+            let s = <String>::deserialize(deserializer)?;
+
+            let s = s.trim_start_matches("0x");
+            let s = <[u8; 32]>::from_hex(s).map_err(serde::de::Error::custom)?;
+
+            Ok(Digest(s))
+        } else {
+            #[derive(::serde::Deserialize)]
+            #[serde(rename = "NewDigest")]
+            struct Value([u8; 32]);
+
+            let value = Value::deserialize(deserializer)?;
+            Ok(Digest(value.0))
+        }
+    }
+}
+
+impl Serialize for Digest {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if serializer.is_human_readable() {
+            format!("{:#x}", self).serialize(serializer)
+        } else {
+            serializer.serialize_newtype_struct("NewDigest", &self.0)
+        }
+    }
+}
+
+impl TryFrom<Vec<u8>> for Digest {
+    type Error = hex::FromHexError;
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        let mut bytes = [0u8; 32];
+        let len = value.len();
+        bytes[..len].copy_from_slice(&value);
+
+        Ok(Digest(bytes))
+    }
```

I think we don't want to silently truncate incoming Vec's one way or the other? You can see [how I did something similar here](https://github.com/agglayer/agglayer/blob/main/crates/agglayer-grpc-types/src/compat/v1/digest.rs)

---

I guess it falls back to this thread: https://github.com/agglayer/provers/pull/57#discussion_r1973916914